tso-server/src/main/java/org/apache/omid/tso/CurrentTimestampOracleImpl.java

```diff
@@ -0,0 +1,171 @@
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+import static org.apache.omid.metrics.MetricsUtils.name;
+
+/**
+ * The Timestamp Oracle that gives monotonically increasing timestamps based on world time
+ */
+@Singleton
+public class CurrentTimestampOracleImpl implements TimestampOracle {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CurrentTimestampOracleImpl.class);
+
+    @VisibleForTesting
+    static class InMemoryTimestampStorage implements TimestampStorage {
+
+        long maxTime = 0;
+
+        @Override
+        public void updateMaxTimestamp(long previousMaxTime, long nextMaxTime) {
+            maxTime = nextMaxTime;
+            LOG.info("Updating max timestamp: (previous:{}, new:{})", previousMaxTime, nextMaxTime);
+        }
+
+        @Override
+        public long getMaxTimestamp() {
+            return maxTime;
+        }
+
+    }
+
+    private class AllocateTimestampBatchTask implements Runnable {
+        long previousMaxTime;
+
+        AllocateTimestampBatchTask(long previousMaxTime) {
+            this.previousMaxTime = previousMaxTime;
+        }
+
+        @Override
+        public void run() {
+            long newMaxTime = (System.currentTimeMillis() + TIMESTAMP_INTERVAL_MS) * MAX_TX_PER_MS;
+            try {
+                storage.updateMaxTimestamp(previousMaxTime, newMaxTime);
+                maxAllocatedTime = newMaxTime;
+                previousMaxTime = newMaxTime;
+            } catch (Throwable e) {
+                panicker.panic("Can't store the new max timestamp", e);
+            }
+        }
+    }
+
+    static final long MAX_TX_PER_MS = 1_000_000; // 1 million
+    private static final long TIMESTAMP_REMAINING_THRESHOLD = 3_000 * MAX_TX_PER_MS; // max number of transactions in 3 seconds
+    private static final long TIMESTAMP_INTERVAL_MS = 10_000; // 10 seconds interval
+
+    private long lastTimestamp;
+    private long maxTimestamp;
+
+    private TimestampStorage storage;
+    private Panicker panicker;
+
+    private long nextAllocationThreshold;
+    private volatile long maxAllocatedTime;
+
+    private Executor executor = Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder().setNameFormat("ts-persist-%d").build());
+
+    private Runnable allocateTimestampsBatchTask;
+
+    @Inject
+    public CurrentTimestampOracleImpl(MetricsRegistry metrics,
+                               TimestampStorage tsStorage,
+                               Panicker panicker) throws IOException {
+
+        this.storage = tsStorage;
+        this.panicker = panicker;
+
+        metrics.gauge(name("tso", "maxTimestamp"), new Gauge<Long>() {
+            @Override
+            public Long getValue() {
+                return maxTimestamp;
+            }
+        });
+
+    }
+
+    @Override
+    public void initialize() throws IOException {
```

Comment:
This is the code for the failover. It allocates the next interval and waits in case the next interval starts in the future. This can occur when the previous TSO already allocated the current interval and we need to wait it will over. The reason we are waiting is to guarantee that the timestamp really representing the world time the transaction was created/committed.